### ISSUE TITLE
Crash after about 12h. Sluggish response when reopened

### ISSUE THREAD

**Issue by kylehase** [thanksmister/wallpanel-android#268]
**Describe the bug**
The app functions normally for hours but will crash after some time. When the app is reopened, the UI is terribly slow. Loading the default dashboard takes 3-5s. Pressing the settings button will show a black screen for a few seconds before loading the settings screen.

On my last test, the crash occurred at around 12hrs, according to MQTT device history in Home Assistant.

Force closing the app doesn't seem to help. A full OS reboot restores normal functionality until another crash.

I'm now testing with MQTT completely disabled in settings to see if this affects stability or timing of the crash.

**To Reproduce**
1. Configure WP to connect to HA instance
2. Wait a few hours and check back

(not sure if the following are required to reproduce the crash, but it's how my instance was configured)
- MQTT client enabled
- MQTT discovery enabled
- Face detection enabled
- Motion detection & QR code disabled
- Camera streaming disabled
- No WP passcode
- WP not configured as Android home button app

**Expected behavior**
The device should be running at all times, turn the screen on when face is detected, respond responsively to user inputs, continue reporting stats to MQTT, respond to MQTT commands.

**Smartphone (please complete the following information):**
 - Device: Google Pixel C tablet (3GB RAM)
 - OS: Android 8.1 build OPM8.190605.005
 - Browser: Stock browser
 - Wallpenal Version: 0.9.4 Build 6

**Additional context**
This is a cleanly wiped Pixel C running the official Google firmware and only Wallpanel app installed. I've disabled "battery optimization" in Android settings for the Wallpanel app to try to prevent Android from killing it. 

I would not consider my HA dashboard heavy, though there are a few custom cards with additional JS files. There are only 12 cards with simple icons and vector graphics, no cameras/video.

**Comment by Shalobee** [comment]
I have been noticing the same thing on my galaxy tab a7 10.4" just to +1 this. I am not using face detection but instead motion detection

**Comment by kylehase** [comment]
@Shalobee are you using MQTT? 

I'm currently testing without MQTT as [this comment](https://github.com/thanksmister/wallpanel-android/issues/192#issuecomment-630985460) from thanksmister implies MQTT may cause issues, though it's from a year ago.

**Comment by thanksmister** [comment]
I don't think this is related to MQTT.  The issue with MQTT happens on devices running Android 9.0 because the MQTT Paho library is running a service without proper permission and this causes the service to crash.  This crashes the service, but does not "hang" the application.  

An application hang when viewing a webpage would more likely be related to the contents of that page hanging the built-in WebView component, which is used for rendering webpages in the application. 

We can test this by using a simple webpage for testing, to see if the app stops working after set amount of time.  Another test would be to not use the camera features, since this increased the application memory and possibly the Android system is closing the app to save resources. 

I am testing the app again on an Android 8 device, but it will take some time to verify the issue.

**Comment by thanksmister** [comment]
I have a crash report for a `0.9.4 Build 2 (9402) Android 6.0.1 Nexus 7 (2012)` related to the bottom sheet not getting displayed but this would be a crash when trying to enter a code for opening the settings.

**Comment by kylehase** [comment]
As @thanksmister suggested, disabling MQTT had no effect. The app crashed multiple times today, was sluggish after re-launching, and only returned to normal operation after OS reboot.

Re-enabling MQTT and disabling all camera functions to continue testing.

**Comment by thanksmister** [comment]
My bet is that the camera features or the webpage you are loading is causing the issues.  Possibly your version of WebView is crashing or not restarting maybe do to some JS or video feature on the page.  The app cannot address this type of crash as its related to the web page rendering and your version of WebView component.

**Comment by kylehase** [comment]
> My bet is that the camera features or the webpage you are loading is causing the issues. 

The app has been stable for over 12 hours after disabling the camera features, so this is very likely the culprit. I was relying on the camera features to wake the screen, as my power button isn't easily accessible. Is this issue unfixable by Wallpanel? I may consider flashing LineageOS 18 (Android 11) if it may help.

**Comment by thanksmister** [comment]
There are a few things we should try.   In WP there are some settings to lower the FPS of the camera, making it less energy intensive because it processes frames slower.   There is also a toggle for hardware acceleration you can try.   If you use the sensor publishing feature, you can monitor your battery charge, it seems that at some point after 12 hours, the system is doing a clean up of apps/tasks due to energy consumption.  This could indicate that your battery level has possibly dropped, you could try dimming your screen or using Android's built-in daydream feature instead of the screensaver in the app, this auto dims your screen to save energy when always on.

**Comment by kylehase** [comment]
Thanks for your continued support.

I re-enabled the camera in face detection mode. Set the FPS to 5 (originally using default of 15). Hardware acceleration was already on, so I left it that way. Disabled WP screensaver and enabled system screensaver.

As for the battery, the tablet is on a smart plug which responds to WP MQTT battery level, turning on at 60% and off at 85%, so at no point was it below 60% except when MQTT was disabled for testing or WP crashed.

**Comment by kylehase** [comment]
Unfortunately this didn't work. After about 12hrs the MQTT reporting became irregular for another 2hrs then finally stopped. The smart plug was in the "on" state during the failure and battery level rising from 73-77% before the crash. WP was no longer running on the tablet at that point and opening WP resulted in sluggish behavior.

**Comment by thanksmister** [comment]
I know this of now help besides information.  I have been running on my Android 8 device for 3 days now with motion detection without issue.   I think the power is related to the behavior you are seeing, at least the app closing.  It could also be that the wifi gets disabled at some point and the page you load freezes.   I have another issue reported that disconnecting wifi causes a similar behavior that I am investigating.

**Comment by thanksmister** [comment]
But I think you have tested this without MQTT and still had the same issue, so this might be unrelated to you, but related to those running Android 8 since it's impacting multiple devices.

**Comment by thanksmister** [comment]
Take your FPS rate down to below 5 which saves CPU.

**Comment by kylehase** [comment]
FPS1 still crashed. I even enabled the Android debug option "Stay awake when charging" and left the tablet charging, in the hopes that this would keep WiFi on and disable other power saving features. The screen was on every time glanced over to check but after ~8hrs WP had closed and the Android home screen was displayed. Restarting WP either lagged or crashed constantly until rebooting the tablet.

At this point I've rooted the device and will either try to periodically kill and restart whichever service may be hanging, or in the worst case hot reboot. I tried killing com.google.android.gms and cameraservice but those didn't seem to have been causing the lag.

**Comment by sygys** [comment]
Wallpanel on my tablet seems to crash also. but only after an android system update.Every update it takes my an hour to get everthing back up. Im not really sure what happens but for some reason android updates overwrite permissions in which the app cant function normally anymore. Things like the camera will stop working properly and the app crashes back to home screen after some time of no use. 

What i do is delete all data from the app reboot the device and set everything back up. mostly this fixes the problem the first time.

Its a pain to do because my samsung tablet is mounted with a secure wallmount which is a pain to get off the wall. Without i cant reboot the tablet. But everytime i setup the app from scratch and reboot the tablet it seems to be working again.

What i did notice now since the last update of android to version 11 with one UI 3.1 is that everytime the tablet wakes up and shows home assistant it gives me connection lost in the bottom left corner and then reconnects. It looks like the app goes in some sort of sleep mode when the screen isnt active for some time. I would like it if the app would keep its connection all the time. So that the info on the screen is more responsive.

One last thing. I noticed some sort of memory leak or something. everytime the app runs for a longer period of time without a restart my home assistant pages start reacting slower and slower over time. When the app is restarted everything is fast again.

Also a last question. Is it possible to set the touch and drag sensitivity? Operating a wall mounted tablet is different then operating it in your hand. the tablet registers drags instead of presses an awful lot. this results in button presses not being executed but merely the UI trying to drag the page. Its hard to keep your finger very still to press a button on a small screen on the wall. I wonder if there is a way to maybe adjust this sensitivity of drag and touch. so that it registers presses easier.

**Comment by thanksmister** [comment]
> everytime the tablet wakes up and shows home assistant it gives me connection lost in the bottom left corner and then reconnects.

This message is because the web page lost connection with the server, the server restarted or has lost it's own connection.  Just losing WiFi connectivity on the device will not show this error, it usually means the server went down.    You can test this by just turning on/off your wifi on the device.  On mine, no message appears, but if I restart the server, then I get this message.

**Comment by thanksmister** [comment]
> One last thing. I noticed some sort of memory leak or something. everytime the app runs for a longer period of time without a restart my home assistant pages start reacting slower and slower over time. When the app is restarted everything is fast again.

Might this be because you are running the camera for long periods of time also?  I thought we established on this ticket that running the camera causes the app to slow on some devices.    A good test would be to see if you can lower the FPS to below 5 or disable the camera to see if the slowdown happens.   

I am still testing the new version of wallpanel on devices running with camera for long periods of time.    Also, if you are using Android 11, there is a possibility that MQTT is causing a crash due to the Android Paho library missing permissions for Android 9+ versions to run background services.     You might try not running MQTT, but running HTTP instead.

**Comment by sygys** [comment]
thanks for the reply i will try these options. 

Im not running mqtt but am using the camera on 15 fps. The tablet i use is indeed a budget one, Samsung A7 2020. I will try to set the fps to below 5

**Comment by thanksmister** [comment]
> thanks for the reply i will try these options.
> 
> Im not running mqtt but am using the camera on 15 fps. The tablet i use is indeed a budget one, Samsung A7 2020. I will try to set the fps to below 5

Maybe check how the battery is doing also, if it is losing power, it may be closing apps or closing the network to save energy.  This may be related to the camera running as well.

**Comment by sygys** [comment]
Battery is a constant 100% with a temperature of 20 degrees celsius with camera on. So i dont think charging/battery is the problem in my case. Ill have to check if the issues  are still there since latest update. could be they are gone now.

**Comment by thanksmister** [comment]
I have a new version which will detect network disconnect and reloads the web page when the network is again connected.  This might solve for some of your issues with the wifi on your device disconnecting.   https://github.com/thanksmister/wallpanel-android/releases/tag/v0.9.4.9

**Comment by Shalobee** [comment]
So not sure if this helps but I may have found the culprit for mine. I am using this for home assistant so its pointing to my home assistant server. I left it open on a different much simpler tab with only a few light buttons on it and it didn't crash. So I started removing cards from my main page until it stopped. It appears to be a custom weather card I was using. I have switched it to the one built in to home assistant and now waiting to see if that fixed it. I also lowered the fps of the camera to 2 but it was still crashing after that so not sure how much of an impact that had.

**Comment by thanksmister** [comment]
Yeah that makes sense since I was also able to produce the issue when navigating around in the HA settings and integration tabs.  Lovelace is extremely heavy to render and in my older devices it fails.  I usually recommend simpler dashboards or using Tileboard for the tablets.

**Comment by spudje** [comment]
I also have this issue. After some hours the app exits/crashes.

I have used the app in the past (on the same hardware) and remember it was quite stable back then. unfortunately not anymore. I'm not using any camera/motion detection features. But I am using MQTT an REST api.

I running on a Lenovo Tab M8 (TB-8505FS) with Android 10.

**Comment by thanksmister** [comment]
There is a known issue with the MQTT library and Android 10 permissions.  This is because the Paho MQTT library is not updated to work with Android 10.  It causes crashes after some time.   Unfortunately, I can’t address this issue until the MQTT project is updated.  It may nave be by the looks of it.

**Comment by thanksmister** [comment]
Try only using the Rest library and not MQTT.

**Comment by sygys** [comment]
@spudje as this project seems to rely on everyone else I recommend to use fully kiosk browser instead. I moved from this app to fully kiosk a while back and I love it. Never had any crashes anymore since. This project here is a great project but it depend to much on other projects that maybe never fixed. As soon as android comes with an update this app might as well be completely broken. I really loved the face recognition software in this app. But I really can't live with the crashes all the time. It's better to have less functions but stable then a wall mounted tablet that has to be unlocked and the app started every time you want to look at it. The whole point of the wall mount and wallpanel is to quickly glance at the screen and get info. Something this app almost never managed for me.

**Comment by thanksmister** [comment]
Unfortunately, some users have taken to disparage the app rather than contribute in a positive way.  This topic is now closed for comments.
